sage.py

```diff
@@ -1588,108 +277,98 @@ def load_data(path, t):
         name = os.path.basename(f)[:-5]
         print(name)
         team_labels.append(name)
-        unparse_ = []
-        
-        unparse_ = parse(readfile(f), [])
-        
-        unparse_ = removeDup(unparse_, t=t)
+
+        parsed_alerts = parse(readfile(f), [])
+
+        parsed_alerts = remove_duplicates(parsed_alerts, gap=t)
         
-        # EXP: Limit alerts by timing is better than limiting volume because each team is on a different scale. 50% alerts for one team end at a diff time than for others
-        start_hours = _s_ # which hour to start from?
-        end_hours = _e_ # which hour to end at?
-        end_time_limit = 3600*end_hours 
-        start_time_limit = 3600*start_hours
+        # EXP: Limit alerts by timing is better than limiting volume because each team is on a different scale.
+        # 50% alerts for one team end at a diff time than for others
+        end_time_limit = 3600 * end_hour       # Which hour to end at?
+        start_time_limit = 3600 * start_hour   # Which hour to start from?
         
-        first_ts = unparse_[0][8]
+        first_ts = parsed_alerts[0][8]
         startTimes.append(first_ts)
-        filtered_unparse = []
-        
-        
-        filtered_unparse = [x for x in unparse_ if (((x[8]-first_ts).total_seconds() <= end_time_limit) \
-                                                and ((x[8]-first_ts).total_seconds() >= start_time_limit))]
-                                                
-        unparse.append(filtered_unparse)
+
+        filtered_alerts = [x for x in parsed_alerts if (((x[8] - first_ts).total_seconds() <= end_time_limit)
+                                                and ((x[8] - first_ts).total_seconds() >= start_time_limit))]
+        team_alerts.append(filtered_alerts)
         
-    return (unparse, team_labels)
+    return team_alerts, team_labels
     
 ## 11
-## Plotting for each team, how much categories are consumed
-def plot_histogram(unparse, team_labels):
+# Plotting for each team, how many categories are consumed
+def plot_histogram(team_alerts, team_labels):
     # Choice of: Suricata category usage or Micro attack stage usage?
     SURICATA_SUMMARY = False
-    cats = {'A Network Trojan was detected': 0, 'Generic Protocol Command Decode': 1, 'Attempted Denial of Service': 2, 'Attempted User Privilege Gain': 3, 'Misc activity': 4, 'Attempted Administrator Privilege Gain': 5, 'access to a potentially vulnerable web application': 6, 'Information Leak': 7, 'Web Application Attack': 8, 'Successful Administrator Privilege Gain': 9, 'Potential Corporate Privacy Violation': 10, 'Detection of a Network Scan': 11, 'Not Suspicious Traffic': 12, 'Potentially Bad Traffic': 13, 'Attempted Information Leak': 14}
+    suricata_categories = {'A Network Trojan was detected': 0, 'Generic Protocol Command Decode': 1, 'Attempted Denial of Service': 2,
```

Comment:
Is this `SURICATA_SUMMARY` still needed? You have mentioned in one of the previous comments that SAGE uses the micro attack stages (Moskal). I have tried running with `SURICATA_SUMMARY = True` and got an exception on the line below (`percentages[tid][suricata_categories[alert[6]]] += 1`: `KeyError: 'Decode of an RPC Query'`). Can it be the case that not all Suricata categories are present?

Reply:
Yes suricata_summary is needed. It was meant to provide two configurations to plot the histograms, which I dont remember anymore, but the code is technically supposed to be functional.

Reply:
And indeed, I think the issue is with the cats = {}. You may leave a comment next to where cats is defined that it needs to be updated

Reply:
I will add a comment then